pkg/muxer/tcp/mux.go

```diff
@@ -326,3 +329,124 @@ func hostSNI(tree *matchersTree, hosts ...string) error {
 
 	return nil
 }
+
+// hostSNIRegexp checks if the SNI Host of the connection matches the matcher host regexp.
+func hostSNIRegexp(tree *matchersTree, templates ...string) error {
+	if len(templates) == 0 {
+		return fmt.Errorf("empty value for \"HostSNIRegexp\" matcher is not allowed")
+	}
+
+	var regexps []*regexp.Regexp
+
+	for _, template := range templates {
+		preparedPattern, err := preparePattern(template)
+		if err != nil {
+			return fmt.Errorf("invalid pattern value for \"HostSNIRegexp\" matcher, %q is not a valid pattern: %w", template, err)
+		}
+
+		regexp, err := regexp.Compile(preparedPattern)
+		if err != nil {
+			return err
+		}
+
+		regexps = append(regexps, regexp)
+	}
+
+	tree.matcher = func(meta ConnData) bool {
+		// FIXME(romain): is it legit ?
+		if meta.serverName == "" {
+			return false
+		}
+
+		for _, regexp := range regexps {
+			if regexp.MatchString(meta.serverName) {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	return nil
+}
+
+// TODO(romain): expose newRouteRegexp in containous/mux fork to get rid of this copied code.
+// preparePattern builds a regexp pattern from the initial user defined expression.
+// this function reuses the code dedicated to host matching of the newRouteRegexp func from the gorilla/mux library.
```

Comment:
I would directly link to the exact point in time and space of the function in github instead.